### PR TITLE
Increases the uWebSockets max backpressure

### LINKED ISSUE #758: Catalog data stream stopped after first return
https://user-images.githubusercontent.com/22226157/109909406-49669c00-7c63-11eb-8938-19a13117a423.mp4

### PR DESCRIPTION
**Pull request by markccchiang**
Closes #758. Increases the uWebSockets max backpressure. According to the [doc](https://unetworking.github.io/uWebSockets.js/generated/interfaces/websocketbehavior.html), `maxBackpressure` is the maximum length of allowed backpressure per socket when PUBLISHING messages. Slow receivers with too high backpressure will be skipped until they catch up or timeout. Also, I add an `OnDrain` handler for when WebSocket backpressure drains. According to the document, we can check the buffered amount here and use this to guide/drive your backpressure throttling.

### CODE REVIEW

**Comment by kswang1029** [comment]
@panchyo0 could you give this fix a try too?

File: src/Main.cc [hunk 1]
```diff
@@ -155,6 +155,17 @@ void OnDisconnect(uWS::WebSocket<false, true>* ws, int code, std::string_view me
...
 }
 
+void OnDrain(uWS::WebSocket<false, true>* ws) {
+    uint32_t session_id = static_cast<PerSocketData*>(ws->getUserData())->session_id;
+    Session* session = sessions[session_id];
+    if (session) {
+        spdlog::debug("Client {} [{}] WebSocket backpressure drains, remaining buffered amount: {} (bytes)", session->GetId(),
+            session->GetAddress(), ws->getBufferedAmount());
+    } else {
+        spdlog::debug("Unknown client WebSocket backpressure drains, remaining buffered amount: {} (bytes)", ws->getBufferedAmount());
```
Comment (confluence) [inline src/Main.cc]: I would change these to "Draining WebSocket backpressure: client {} [{}]. Remaining buffered amount: {} (bytes)."

**Comment by kswang1029** [comment]
Tested with a 1M source catalog and data stream works fine. Other stream services (histogram, spectral profile etc) work too.

File: src/Session.cc [hunk 2]
```diff
@@ -1571,6 +1571,12 @@ void Session::SendEvent(CARTA::EventType event_type, uint32_t event_id, const go
         std::pair<std::vector<char>, bool> msg;
         if (_connected) {
             while (_out_msgs.try_pop(msg)) {
+                auto expected_buffered_amount = msg.first.size() + _socket->getBufferedAmount();
+                if (expected_buffered_amount > MAX_BACKPRESSURE) {
+                    spdlog::warn(
+                        "Client {} [{}] WebSocket buffered amount {} (bytes) exceeds the max backpressure! May losing some messages..",
```
Comment (confluence) [inline src/Session.cc]: Change "May losing" to "May lose"; remove extra trailing full stop.

Comment (confluence) [reply src/Session.cc]: Also change "Client {} [{}] WebSocket buffered amount {} (bytes) exceeds the max backpressure!" to "Exceeded maximum backpressure: client {} [{}]. Buffered amount: {} (bytes)."

**Review by confluence — CHANGES_REQUESTED** [review]

**Review by confluence — APPROVED** [review]

**Comment by markccchiang** [comment]
@confluence Thanks for the corrections!

**Review by kswang1029 — APPROVED** [review]

**Pull request merged**